### ISSUE TITLE
Neutral bone - error on MSFS sim build - Read me

### ISSUE THREAD

**Issue by ronh991** [ronh991/glTF-Blender-IO-MSFS#17]
### Current Behavior

If you have vertices with a armature weighting of zero in vertex groups for armature animations.  Blender/Khronos will add a neutral bone to your glTF on export - this will cause an error in the MSFS build, talking about a neutral_bone duplicate naming. 

Zero weighting in weight paint causes this issue.

https://github.com/KhronosGroup/glTF-Blender-IO/issues/2303

### Expected Behavior

No neutral_bone in glTF.

### Steps To Reproduce

Bad vertex group happens in parenting with automatic weights.

### Environment

```markdown
- OS:win 10
- Blender:3.6 and 4.2
- glTF-Blender-IO-MSFS: any that work with 3.6 and 4.2
```

**Comment by ronh991** [comment]
EDIT:  *********** DO NOT USE THIS FIX AS THERE IS A COMMIT THAT fixes the neutral bone issue. ******************************


For those brave enough to modify Khronos python code. MAKE BACKUPS!!


The gltf2_blender_gather_primatives_extract.py in the 
Program Files\Blender Foundation\Blender 4.2\4.2\scripts\addons_core\io_scene_gltf2\blender\exp
folder

replace the __get_bone_data function

I just added the code pertaining to pre_meshname variable

```
    def __get_bone_data(self):

        self.need_neutral_bone = False
        min_influence = 0.0001

        joint_name_to_index = {joint.name: index for index, joint in enumerate(self.skin.joints)}
        group_to_joint = [joint_name_to_index.get(g.name) for g in self.blender_vertex_groups]

        # List of (joint, weight) pairs for each vert
        self.vert_bones = []
        max_num_influences = 0

        prev_meshname = ""
        for vertex in self.blender_mesh.vertices:
            bones = []
            if vertex.groups:
                for group_element in vertex.groups:
                    weight = group_element.weight
                    if weight <= min_influence:
                        continue
                    try:
                        joint = group_to_joint[group_element.group]
                    except Exception:
                        continue
                    if joint is None:
                        continue
                    bones.append((joint, weight))
            bones.sort(key=lambda x: x[1], reverse=True)
            if not bones:
                # Is not assign to any bone
                bones = ((len(self.skin.joints), 1.0),)  # Assign to a joint that will be created later
                self.need_neutral_bone = True
                if prev_meshname != self.blender_mesh.name:
                    print("***** Khronos Core-MSFS Warning neutral_bone added *****", vertex, self.blender_mesh.name)
                    prev_meshname = self.blender_mesh.name
            self.vert_bones.append(bones)
            if len(bones) > max_num_influences:
                max_num_influences = len(bones)

        # How many joint sets do we need? 1 set = 4 influences
        self.num_joint_sets = (max_num_influences + 3) // 4
```

The system console window will have something like this

```
13:08:25 | INFO: Primitives created: 8
13:08:25 | INFO: Extracting primitive: Mesh.197
***** Khronos Core-MSFS Warning neutral_bone added ***** <bpy_struct, MeshVertex at 0x000001EFA7AED708> Mesh.197
13:08:25 | INFO: Primitives created: 1
13:08:25 | INFO: Extracting primitive: Mesh.196
```

**Comment by erasam** [comment]
I have modified the code accordingly and I see the warning being printed but I have always the duplicate name error while compiling the project in MSFS

**Comment by ronh991** [comment]
What is the error? and ensure you don't have the same name object/mesh/bone in the blend file

**Comment by ronh991** [comment]
This fix is no longer required - there is a commit that fixes the neutral bone issue.

**Comment by erasam** [comment]
I still have the error message "Node without Unique ID while using the unique_id extension, nodes[224] : 'neutral_bone'" and there are two instances of neutral_bone in the generated gltf files as I have two armatures (the first as a copy of the second)

I thought no neutral_bone should be generated at all with the latest version of your exporter

**Comment by ronh991** [comment]
AH yes - are you using 3.6 or 4.2 Blender. I forgot to add the unique id info.  Sorry will fix soon.

**Comment by erasam** [comment]
I'm using Blender 3.6 LTS

**Comment by erasam** [comment]
In the exported file is it planned to have no neutral_bone then ?

**Comment by ronh991** [comment]
Yes It will have a neutral bone, just needs to be renamed  this is a Khronos thing - that they added, not ASOBO, but ASOBO needs to handle it.  They have fixed this in the 2024 exporter but not the 2020 exporter.

To be sure - you are using the 2020 exporter with Blender 3.6.x

**Comment by erasam** [comment]
Confirm, I'm using Blender 3.6 and exporter for MSFS2020

**Comment by ronh991** [comment]
Try 1.6.12 release

https://github.com/ronh991/glTF-Blender-IO-MSFS/releases/tag/v1.6.12.0

**Comment by erasam** [comment]
No more errors now !

**Comment by ronh991** [comment]
Thanks, appreciate you testing an reporting errors/issues Thanks

**Comment by erasam** [comment]
Unfortunately I'm suffering this "Base Color not being used in export" as I have some standard black texture shown in MSFS as white (working fine with the standard exporter)

**Comment by ronh991** [comment]
You will have to show me the exact error message.  And Possibly give me an example blend file that makes that error.  I have no idea what that is.

**Comment by erasam** [comment]
I read this in the list of issues (inside the documentation)
"#11 Base Color not being used in export - fixed - bug introduced"
When I export the mesh in the attached blend file the color shown in MSFS is full white instead of the black configured in Blender changing the base color

[Pedal.zip](https://github.com/user-attachments/files/18862690/Pedal.zip)

**Comment by ronh991** [comment]
I don't understand what you are doing - It's not an MSFS material

![Image](https://github.com/user-attachments/assets/ded621e4-2318-4fec-a7ce-25768162cafe)

**Comment by erasam** [comment]
> I don't understand what you are doing - It's not an MSFS material
> 
> ![Image](https://github.com/user-attachments/assets/ded621e4-2318-4fec-a7ce-25768162cafe)

You are right, it's a simple standard Blender material. With the Asobo exporter rel. 1.3.3 is anyway rendered correctly in MSFS as black, with your version is shown as plain white

**Comment by ronh991** [comment]
You are not understanding how the exporter works - there are numerous msfs_* variables that need to be set - the exporter uses those variables to set the gltf parameters.  Since you did not use an MSFS material the default base color is white.  So it uses that for the sim.

You were just lucky it worked in the past.

I have said over and over again ... "that you cannot use the surface panel or manipulate the nodes directly - you MUST use the MSFS panel to set the material"  

![Image](https://github.com/user-attachments/assets/0b62d0c3-5e75-4c3e-8053-6d44aeda8e6d)

**Comment by erasam** [comment]
Ok, got it

**Comment by erasam** [comment]
I was worried to redo all my Blender standard materials from scratch when I discovered the great "Pre-Migrate Legacy Data" tool: a real lifesaver !

The only issue I still have is my windshields are rendered blurred as in the pic (they are rendered properly with the official Asobo Exporter)

![Image](https://github.com/user-attachments/assets/e062c538-5008-4536-9679-ec638b175136)

**Comment by erasam** [comment]
I found the issue: it was the roughness not being 0, probably the Asobo exporter forces it automatically
